include/nbl/asset/filters/kernels/WeightFunctions.h

```diff
@@ -0,0 +1,308 @@
+		}
+		return 0.0;
+	}
+
+private:
+	static inline constexpr double b = double(B::num) / double(B::den);
+	static inline constexpr double c = double(C::num) / double(C::den);
+	static inline constexpr double p0 = (6.0 - 2.0 * b) / 6.0;
+	static inline constexpr double p2 = (-18.0 + 12.0 * b + 6.0 * c) / 6.0;
+	static inline constexpr double p3 = (12.0 - 9.0 * b - 6.0 * c) / 6.0;
+	static inline constexpr double q0 = (8.0 * b + 24.0 * c) / 6.0;
+	static inline constexpr double q1 = (-12.0 * b - 48.0 * c) / 6.0;
+	static inline constexpr double q2 = (6.0 * b + 30.0 * c) / 6.0;
+	static inline constexpr double q3 = (-b - 6.0 * c) / 6.0;
+};
+
+// Kaiser filter, basically a windowed sinc.
+struct SKaiserFunction
+{
+	constexpr static inline float min_support = -3.f;
+	constexpr static inline float max_support = +3.f;
+	constexpr static inline uint32_t k_smoothness = 3; // TODO(achal): Figure out what this should be
+	// important constant, do not touch, do not tweak
+	static inline constexpr float alpha = 3.f;
+
+	template <int32_t derivative = 0>
+	static inline double weight(float x)
+	{
+		if (x >= min_support && x < max_support)
+		{
+			const float absMinSupport = core::abs(min_support);
+
+			if constexpr (derivative == 0)
+			{
+				const auto PI = core::PI<double>();
+				return core::sinc(x * PI) * core::KaiserWindow(x, alpha, absMinSupport);
+			}
+			else if constexpr (derivative == 1)
+			{
+				const auto PIx = core::PI<double>() * x;
+				double f = core::sinc(PIx);
+				double df = core::PI<double>() * core::d_sinc(PIx);
+				double g = core::KaiserWindow(x, alpha, absMinSupport);
+				double dg = core::d_KaiserWindow(x, alpha, absMinSupport);
+				return df * g + f * dg;
+			}
+			else
+			{
+				static_assert(false, "TODO");
+				return core::nan<double>();
+			}
+		}
+		return 0.0;
+	}
+};
+
+// This is the interface for 1D weight functions that can be used to create a `CFloatingPointSeparableImageFilterKernel`.
+// Current implementations of this interface are:
+// - `CWeightFunction1D`
+// - `CChannelIndependentWeightFunction1D`
+// - `CConvolutionWeightFunction1D`
+template<typename T>
+concept KernelWeightFunction1D = requires(T t, const float x, const uint8_t channel)
+{
+	{ t.weight(x, channel) }	-> std::same_as<double>;
+	{ t.getMinSupport() }		-> std::same_as<float>;
+	{ t.getMaxSupport() }		-> std::same_as<float>;
+	{ t.stretch(x) }			-> std::same_as<void>;
+	{ t.scale(x) }				-> std::same_as<void>;
+	{ t.stretchAndScale(x) }	-> std::same_as<void>;
+};
+
+template <typename Function1D, int32_t derivative = 0>
+class CWeightFunction1D /*final*/ // TODO(achal): Cannot make it final just yet because `weight_function_value_type` inherits from this.
+{
+	static_assert(derivative <= Function1D::k_smoothness);
+public:
+	constexpr static inline uint32_t k_smoothness = Function1D::k_smoothness-derivative;
+	constexpr static inline float k_energy[4] = { 0.f, 0.f, 0.f, 0.f }; // TODO(achal): Implement.
```

Comment:
link me the discord convo about this

Reply:
Also due to https://github.com/Devsh-Graphics-Programming/Nabla/pull/435/files#r1196619001 it should be single valued and thats it.